design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -196,35 +201,35 @@ One stream (MIGRATION_EVENTS) and one table (SCHEMA_VERSION).
 The stream and table require topics unique for the cluster. In this case, these topics names will use the same convention as the processing log. It's not going
 to be an internal topic because it will be a user topic. The topic name is: `{clusterID}ksql_{StreamOrTableName}`.
 
-The user can specify a different name for the SCHEMA_VERSION table. This can be done through the tool configuration file. See `Configurations` for more details.
+The user can specify a different name for the MIGRATION_EVENTS stream and SCHEMA_VERSION table. This can be done through the tool configuration file. See `Configurations` for more details.
 
 The `MIGRATION_EVENTS` stream will keep track of every migration change (including undo changes). This will contain the history of changes the user has done. Also, this stream
 will contain a key to the current version of the schema (specified as `CURRENT` version). Every time a new migration or undo happens, the tool will insert a new event with the
 `CURRENT` key pointing to the current version.
 
 This is the `CREATE` statement for the `MIGRATION_EVENTS` stream:
 ```sql
-CREATE STREAM migration_events (
+CREATE STREAM migration_events ( /* name will be configurable */
   version_key  STRING KEY,
   version      STRING,
   name         STRING,
   state        STRING,  
   checksum     STRING,
-  started_on   STRING,
-  completed_on STRING,
+  started_on   STRING, /* may use the new TIMESTAMP type, depending on availability at time of release */
+  completed_on STRING, /* may use the new TIMESTAMP type, depending on availability at time of release */
   previous     STRING
 ) WITH (  
-  KAFKA_TOPIC='default_ksql_migration_events',
+  KAFKA_TOPIC='default_ksql_migration_events', /* topic name will be configurable */
   VALUE_FORMAT='JSON',
   PARTITIONS=1,
-  REPLICAS=1
+  REPLICAS=1 /* value will be configurable */
 );
 ```
 
-The `version_key` column has the version of the migration applied or undone. Special values of the `version_key`, `CURRENT` and `LATEST` will be reserved for internal purposes.
+The `version_key` column has the version of the migration applied or undone. A special value of the `version_key`, `CURRENT`, will be reserved for internal purposes.
 The `version` column has the version of the migration applied.
 The `name` column has the name of the migration.
-The `state` column has the state of the migration process. It can be any of `Pending`, `Running`, `Migrated`, `Error`, `Undone`.
```

Comment:
What was the intended difference between `Pending` and `Running`? I've removed one for now to simplify the status flows (see "Migration state transitions and potential race conditions" below).

Reply:
My understanding was that `Pending` represents migrations which have been defined but not yet applied, where `Running` represents a single in-progress migration.

Reply:
I see the confusion. As @colinhicks mentioned, the `Pending` represents migrations not applied yet. However, this `Pending` status will not form part of the migration schema stream/table because those migrations hasn't run yet. The `Pending` is just a status that will be displayed by `ksql-migrations info` command, which looks at current migrations (in the schema table) and available migrations (in the local directory).

Reply:
OK, I will update the example flows below to use `Running` instead of `Pending`, and still leave `Pending` as a possible status, the only use for which is to show in the output of `info` for migrations which have yet to be applied. Thanks for clarifying!

Reply:
That's an interesting idea. It sounds like we're in agreement that the three states we need internally (in the metadata stream/table) are `running`, `complete`, and `error` (actual names subject to change) and the open discussion is about what states the user should see. The user states must include `running`, `complete`, and `pending` (not yet started), where again the actual names are subject to change. Is it useful for the user to see `error` (meaning a migration was attempted but failed)? I think the answer is yes because there should have a way to see the error message associated with a failure other than the output of the `apply` command itself. I'm imagining another column in the output of `info` called `failure reason` or similar, so we can save failures and show them to the user (rather than requiring them to re-run the `apply` command to remember why the command failed). This column will have to be added to the metadata stream/table as well. WDYT?

Reply:
That sounds good to me.

One thing I'd like to emphasize still, is that the output `info` should only contain one line per version, whereas the migration_events captures a whole history of migration actions, hence each version could have multiple records. I'd want to point out that `info` does not necessarily need to show the whole history, but only the latest snapshot of the migration status per version.

Reply:
We are in agreement here. `info` will only print one line per version (as discovered in the `/migrations` directory). The line printed will be the latest version, as pulled from the metadata table. `/info` will not consume from the metadata stream at all.